### PR TITLE
Remove Suboptimal Code

### PR DESCRIPTION
**Pull request by LunarLambda**
Following the killing of George Floyd by police on May 25th, 2020, people of color across the US have been risking their lives in protests against police brutality and deeply racist system.

Police have been known to threaten, assault, and even ***kill*** protesters, identifying them using information gathered from social media as well as facial recognition algorithms.

The highly questionable ethics of releasing a tool that could be used to subvert anonymization under *any* circumstance aside, releasing this right now is not only incredibly dangerous, but can and **will** get people killed.

I ask you to permanently delete this repository, and everything associated with it.
#BlackLivesMatter

### CODE REVIEW

**Comment by Demonstrandum** [comment]
This technology already exists and will continue to exists regardless of this repository.  I think it is being a bit paranoid and unhelpful to ask someone to delete their personal project, incase *American police* decide to use it.

**Comment by LunarLambda** [comment]
Did you miss the part where I said that this technology is ethically incredibly questionable *regardless* of circumstance or are you just being dishonest.

It already existing doesn't make it any less wrong

**Comment by Demonstrandum** [comment]
This technology on its own is not wrong.  Deleting this project changes nothing.  That was my point.

**Comment by LunarLambda** [comment]
It absolutely is wrong. Anything that can be used to subvert anonymization is wrong, for the same reasons that outlawing encryption would be wrong

Just because it's notperfect doesn't mean it can't be a tool for evil. Deleting this project and others like it may not fix the status quo, but keeping them up will undoubtedly make it worse. For everyone.

**Comment by Diatrus** [comment]
Believing this gentleman is the first to discover and implement this is incredibly naïve and honestly demonstrates an extremely narrow and simpleminded worldview. It’s fine to question the ethics of the use of this software, but to be so immature as to act as if this is groundbreaking to governments is silly at best, counterproductive and harmful at worst.

**Comment by LunarLambda** [comment]
I do not know how to explain to you that you should care about other people

**Comment by Diatrus** [comment]
I don't know how to explain to you that censoring open source implementations of a technology doesn't help get rid of the obvious use of close sourced implementations by governments.

**Comment by LunarLambda** [comment]
It helps prevent the people who are not the government but support the government (like police, or people who just *really* like the government and *really* dislike people that don't like the government) from using it

As programmers we have a lot of power, and, one would hope, a moral obligation to use that power for good, not for evil.
But that would, I don't know, require caring about others, I guess

**Comment by dkter** [comment]
It speaks to a lot that the author of this project decided to spend their time (on Juneteenth, nonetheless) creating a tool that could easily be used to help the police during these protests. It's insensitive and irresponsible, given what's going on.

Keep in mind the core technology behind this project already existed, and the author decided to create a simple frontend in which anyone, without any technical knowledge whatsoever, can click "Open in Colab", upload an image, and begin uncensoring people's faces. If their goal was to alert people that this technology was possible, there are better ways of doing that than publishing a tool that is as easy as possible for anyone to use.

Either way, this project has already made its way around the news. The responsible thing to do now is to take it down.

**Comment by Whanos** [comment]
Like I've said before - a well funded government team could easily make a system similar to this - If not better due to them having a huge training set of faces (think about the ID images the government has!). Deleting this repo will solve nothing.

**Comment by dkter** [comment]
@Whanos 

>It helps prevent the people who are not the government but support the government (like police, or people who just really like the government and really dislike people that don't like the government) from using it

Not to mention, why on earth would you choose to make *this* right now?

**Comment by tg-bomze** [comment]
I am writing in Russian through Google Translator, so I'm sorry for the errors in the text. This neural network originates from 2018. Nvidia company launches StyleGAN, which opens up great opportunities for generating faces. Duke University comrades rewrote it for working with pixel images. I just made a simple and intuitive interface so that each of my readers can experiment. What questions can be for me if all this has been open source for a long time?

Treat this as a warning that it is now possible to recognize facial features even from a low resolution image. It is impossible to restore the original face (you probably saw examples), since the information inside one pixel is negligible.

Like any other technology, this can be used for different purposes. May be beneficial or harmful. In any case, if there is no sharing, then there will be no progress.

**Comment by ghost** [comment]
cope

**Comment by SurealCereal** [comment]
> cope

Probably not taught in the safe spaces

**Comment by dkter** [comment]
ah, safe space. haven't heard that one since 2017 or so. going for the throwback are we?  
you know this isn't about your God-given right to say the N word or whatever. it's bigger than that. small-scale police departments without huge forensics teams are going to use this to subvert anonymization.  
if this had come out at a different time, it would have been a little different. as it stands, considering the consequences of your actions is pretty important for someone working in the tech field.

**Comment by ghost** [comment]
This project has been forked 107 times, this code (unless you would fake DMCA requests) will never be deleted.

**Comment by ghost** [comment]
@dkter @LunarLambda I would advise you to actually read the paper regarding this software and the process itself on how it de-pixelizes faces. 

It doesn't regain the lost information from the image before, it _creates_ a new face that fits the original pixelization. There's a chance of it being correct, that chance being less than getting struck by lightning in a 50 foot bunker dug 100 feet down.

By that, I'm saying that the chance this random face matches a real face is improbable.

**Comment by LunarLambda** [comment]
I will happily reiterate it that it doesn't need to be accurate, it doesn't
even need to be CLOSE to accurate. It simply needs to convince a jury, and
history shows that a lot less than this is needed to accomplish that

**Comment by ghost** [comment]
How on Earth would this ever pass in a court of law? The prosecution could run a MILLION tests and get something completely different **everytime** than the defendant's face. Not only that, but all it would take is a couple of similar runs through the software and get the evidence thrown out entirely by a lawyer with any knowledge of computer science.

If that's all the evidence you have, the case would get thrown out the door.

**Comment by LunarLambda** [comment]
Here's a gentle reminder that cops have been forging evidence on an
as-needed basis for years

https://m.huffpost.com/us/entry/us_5a72543fe4b09a544b564f2a?guccounter=1&guce_referrer=aHR0cHM6Ly9kdWNrZHVja2dvLmNvbS8&guce_referrer_sig=AQAAAKNqqPZm3PlGHbEHqrQ5EGAczHvxMCOhJR9La-OXKjsxO08lcUKUgph8LV0IzLx6fjIAt_Eq-nTVTdUnA9IJjuYAGhdVJ_8KePjbioWtUnd5D7L6TIJ_fPnDBGPEOFENz3xz7YAnOq5HDXKSPxWFTaEdkiYZ6pZHqQxzg9d8UOIv

**Comment by ghost** [comment]
And that's the same thing...how?

It's horrible that police officers forge fake evidence, however, your article was an act of pure hatred to for police to do something like that. It would be questionable at best if an officer ran a face through the Depixelizer and somehow got a "warrant" for the arrest. 

Like I said before, you could generate millions of different faces on the same picture and still get a random person each time other than a regular person. Consider the fact that the cop wouldn't even know actual information about the suspect except for the pixelized face and a fake de-pixelization. Getting a warrant in a court, even FISA, would be impossible.

Not only that, but I would be more afraid of cops ripping off my mask and pepper-spraying me in the face. Your argument that police officers would forge evidence like this and go through all the trouble to arrest...who? What would be the point? I fail to see a logical way that a police officer would go through the time using this software (that they would have to know about offhand) and bring it to a court to furnish a fake warrant.

I really do understand your concern but there's no logical basis that any member of the court, law, or federal would use this specific piece of software to falsely imprison someone.
